src/game/GameRoom.ts

```diff
@@ -0,0 +1,144 @@
+    this.gameField.despawn(player.x, player.y);
+    this.players.delete(client.serverId);
+    logger.info(`Player ${client.serverId} disconnected`, code, reason);
+
+    this.broadcast(new LeaveEvent(client.serverId));
+  }
+
+  private registerJoin(client: Client, name: string) {
+    const player = new Player(client.serverId, name);
+    this.players.set(client.serverId, player);
+    player.spawn(this.gameField);
+
+    client.send(new Message(0, { gameField: this.gameField.safeExport(), self: client.serverId }, "JOIN"), true);
+    this.broadcast(new JoinEvent(player), client);
+    client.on("message", message => this.registerMessage(client, message));
+  }
+
+  private registerMessage(client: Client, message: Message) {
+    const { data, type } = message;
+    logger.debug(`New message from ${client.serverId}`, data, type);
+
+    switch(type) {
+      case "MOVE":
+        this.moveMessage(client, data);
+        break;
+      default:
+        client.send(new Message(0, {}, "INVALID_MESSAGE_TYPE"), true);
+        break;
+    }
+  }
+
+  private moveMessage(client: Client, data: Data) {
+    const player = this.players.get(client.serverId);
+    const direction: MoveDirection = data?.direction;
+    const moveOffset = moveMap[direction];
+    if (!direction || !moveOffset)
+      return client.send(new Message(0, { reason: "Invalid or missing direction" }, "INVALID_DATA"), true);
+
+    if (!player) {
+      logger.error(`Client ${client.serverId} has no player object`);
+      return client.disconnect(1011);
+    }
+
+    const isMoved = this.gameField.move(player, moveOffset);
+    if (isMoved)
+      this.broadcast(new MoveEvent(player));
+  }
+
+  private requestName(client: Client) {
+    const room = this;
+
+    client.send(new Message(0, {}, "NAME_REQUEST"), true);
+    client.on("message", function nameRequestHandler({ data, type }) {
+      if (type != "NAME_REQUEST")
+        return client.send(new Message(0, { reason: "Incorrect message type" }, "INVALID_NAME"), true);
+
+      const name: string = data?.name;
+      if (!name || !Player.validateName(name))
+        return client.send(new Message(0, { reason: "Invalid or missing name" }, "INVALID_NAME"), true);
+
+      if (!room.isNameAvailable(name))
+        return client.send(new Message(0, { reason: "Name already taken" }, "INVALID_NAME")), true;
+
+      if(room.players.size == playersPerRoom)
+        return client.send(new Message(0, {}, "GAME_ROOM_IS_FULL"), true);
+
+      client.removeListener("message", nameRequestHandler);
+      room.emit("join", client, name);
+    });
+  }
+
+  public broadcast(event: GameEvent, author?: Client) {
+    for (const [ playerId ] of this.players) {
+      if (author && author.serverId == playerId) continue;
+      this.clients.get(playerId)?.send(new Message(0, event, "GAME_EVENT"), true);
+    }
+  }
+
+  public isNameAvailable(name: string) {
+    let available = true;
+    this.players.forEach(player => {
```

Comment:
We can call `.values()` instead and iterate through it. I chose `forEach` because this way the code is more readable and the performance loss in this place is insignificant (maximum size of Map is 16).

Reply:
Had run some tests, `forEach` has the best performance.